### ISSUE TITLE
Review changes required to connect to MS SQL ePJS database in a non-persistent manner?

### ISSUE THREAD

**Issue by reddog** [healthlocker/healthlocker#1090]
This issue falls out of the work done at https://github.com/healthlocker/healthlocker/issues/1078.

The connection to the MS SQL ePJS extract database is currently set to be persistent.  This is achieved by all database connections performing a 

```
SELECT 1
```

database query every second.  There are currently 10 database connections opened.

These *keep-alive* queries are causing a 2-3% load on the database server (plus network traffic, etc).

What is required is:

- a review of the changes required to establish non-persistent database connections (changes to the app startup process)
- a review of what (if any) changes would be required for the app to re-connect at the point that an ePJS connection is required but those connections have been terminated (changes to the app in-use process)

**Comment by reddog** [comment]
@katbow discovered an article which talks about `ecto` and Postgres keep-alive issues, see https://github.com/elixir-ecto/ecto/issues/195

This *might* suggest that one approach could be:

1.  establish non-persistent database connections (how - what configuration settings for MS SQL?)
2.  at time of database querying, if a `TCP closed` error (or whatever the correct MS SQL error is) is returned when attempting the query, then re-establish database connectivity and re-run the query.

**Comment by katyedwards** [comment]
Propose initial 2 hour spike to think/ research.  How does that sound?

**Comment by katyedwards** [comment]
FYI @katbow it seems as though we may _have_ to remove the persistent/ 'keep alive' database connections so if you can look into it with that in mind that would be helpful - or at least get some idea of the difference in time required to reduce vs remove

**Comment by katbow** [comment]
Do we know if they are using SQL Server 2014 or later? If so then it *should* be ok to [stop the ping entirely](https://docs.microsoft.com/en-us/sql/connect/odbc/windows/connection-resiliency-in-the-windows-odbc-driver).

**Comment by reddog** [comment]
I believe it's not as recent as SQL Server 2014.

**Comment by katyedwards** [comment]
Have dug out of an email a comment from them:
_In SQL Server 2005, resources are made available at the server level rather than the database level. So all of the databases on a server will consume resources from the same pool._

So I assume 2005...

**Comment by katbow** [comment]
okay, do we have any details of the peak load at the busiest times? And where the monitoring tools show where the bottle neck is?

**Comment by katyedwards** [comment]
Does this help? However the ePJS system has typically 1200 concurrent users running upwards of 800 batches per second.

I'll forward you an email too with some more detail...

**Comment by reddog** [comment]
@katbow No, we have none of that - I don't think that is the issue at all.

What we do know is that the HealthLocker app is opening 10 persistent connections that are performing a `SELECT 1` query each once every second.  This is what's due to be addressed.

Non-persistent connections could mean that a database connection has to be opened at the time of the database query (if the connection has died) which would mean a slight pause whilst database connectivity is re-established.

Does the current implementation handle cases where the MS SQL Server goes away unexpectedly? (I think this has been mentioned elsewhere).  This is a similar type of problem to what's being addressed here (how to recover from a connection being lost/dropped).

**Comment by reddog** [comment]
@katbow It sounds like it would be useful to have a quick call on this.  Would you be free at some point for a chat?

**Comment by katbow** [comment]
> What we do know is that the HealthLocker app is opening 10 persistent connections that are performing a SELECT 1 query each once every second. This is what's due to be addressed.

I understand that, but my question was to try to understand if there are any other factors contributing to this. 2-3% load doesn't sound like something that should significantly reduce performance. I see that there is the correlation, but I am wondering about underlying issues. I can stop the pinging and have a solution for this part, but I also want to ensure that this isn't a bandaid solution and does not cause other issues in the application (such as the database disconnecting).

I am currently looking at how lost connections are handled with Ecto/the mssql adapter, and options to recover a connection once it's been lost.

**Comment by katyedwards** [comment]
@katbow from what I have understood, the situation they are in is that 2-3% does have an impact on performance, but they are mid-way through a project to move ePJS so a band-aid may be all we need at this stage while we wait for them to complete their project which will properly address the problem.

**Comment by reddog** [comment]
@katbow Thanks for that.  Do let me know if you have time for a chat on this.

I think the point that's being made by the ePJS team is that the 2-3% load isn't really *necessary* (as far as ePJS is concerned) and is putting additional load (albeit by itself small) on a critical, already heavily loaded server.

I think the underlying issues are **totally** out of our hands - that's with the ePJS DB admin team.

Certainly, irrespective of the keep-alive queries, HealthLocker should be able to recover from instances where database connections are lost.  If that is already in place, or when it is in place (if not already), then that would cover both unexpected disconnections as well as removing the constant keep-alive pings. 

Hence the designed-for expectation would be that we *may well not* have ePJS connectivity at the time of querying, so we should check if that's the case when we do a query and re-connect if required.  I think that's the opposite of the current design where it is *expected* that we *do have* connectivity, and then possibly can't recover from a situation when we have a dropped connection.

**Comment by reddog** [comment]
@katyedwards The fix should make HealthLocker more resilient in other cases, not just remove this 2-3% load - it shouldn't be just a band-aid fix (there may be other instances where HealthLocker will fail at the moment if the ePJS database were to unexpectedly go away).

**Comment by reddog** [comment]
I have slightly renamed this issue and its requirements to better address that this issue should handle expected and unexpected ePJS database disconnects as opposed to changing from establishing persistent connections to using transient connections.

**Comment by katbow** [comment]
Note for this one which I forgot to put on earlier. I left it for the time being, as I didn't feel like I was making enough progress at the time. I noticed @reddog 's [comment on this issue](https://github.com/findmypast-oss/mssqlex/issues/1#issuecomment-331123282), which will hopefully provide us some insight into our options.

**Comment by reddog** [comment]
@katbow Thanks very much for your update.  I wasn't sure exactly why this issue was left at the time - thanks for giving some insight to that.

Unfortunately we've had no reply to my comment elsewhere, but hopefully we can find solutions put in place by others / patterns fairly easily.

**Comment by iteles** [comment]
@reddog @katyedwards With an issue like this where either specialist knowledge or continuous problem solving is required in a specific are of expertise, I would strongly suggest that in the absence of availability from a specialist, we bring someone in for @katbow to pair with. We have often found this to be a more fruitful means to more complex or unknown debugging, as we benefit both from a fresh perspective as well as a having a team together to discuss different ideas.

In this case, I would propose Daniel - who MV has already worked with recently - work with Katherine on this as he is a particularly skilled problem solver with _excellent_ Elixir knowledge; he is available from Monday or Tuesday onwards (dependent on LDMW). @katyedwards Happy to discuss and agree the particulars, but I would recommend that in light of Katherine’s initial research into this issue and the subject at hand, this would be our most efficient route forward.

**Comment by reddog** [comment]
@iteles Thanks for your input.  Maybe I have been doing some "pairing" already on this issue?  I have updated with findings from elsewhere as well as requesting calls to ensure we're tackling those aspects that are within our grasp.  I am certainly happy to continue to pair on this in order to obtain quick and effective resolution towards a full SLaM signoff.

I have just finished speaking with Garry and I think we have an interim solution that we can work towards.

**Comment by iteles** [comment]
@reddog Thanks for this. @katbow can certainly carry out #1095, we can revisit this at a later date if required 👍

**Comment by reddog** [comment]
@iteles Thanks for that.  As previously expressed (above and elsewhere), I am very happy to be part of a pair to help resolve issues as effectively as possible and would be happy to be taken up on that offer.

I will certainly speak with @katyedwards on Monday regarding your suggestion that we bring in Daniel.

**Comment by iteles** [comment]
@reddog Thank you for this.

In terms of pairing, I apologise for not adequately conveying our thoughts on this.
You and Katherine have done an excellent job of finding possible solutions which you have both documented here and in #1095.

Where we believe Daniel or an experienced Elixir developer will be an essential pair for Katherine is in the _implementation_ of these solutions in Elixir specifically and the troubleshooting of these. It is here that the debugging and bouncing of ideas comes into full effect.

**Comment by Danwhy** [comment]
It looks like this is something that should be, and usually is, handled by the database adapters (eg. https://groups.google.com/forum/#!topic/elixir-lang-talk/ljDAIh-nLxA). It might be worth looking into/testing whether the adapter we're using handles this, or if it's something we need to implement.
